README.md
# Popup

---

[![spm package](http://spmjs.io/badge/arale-popup)](http://spmjs.io/package/arale-popup)
[![Build Status](https://img.shields.io/travis/aralejs/popup.svg?style=flat)](https://travis-ci.org/aralejs/popup)
[![Coverage Status](https://img.shields.io/coveralls/aralejs/popup.svg?style=flat)](https://coveralls.io/r/aralejs/popup)

Popup 是可触发的浮层组件。封装了当用户点击、hover、focus 到指定元素时，可以触发另一个浮层显示的行为。

---

Popup 继承自 [overlay](http://aralejs.org/overlay/)，可使用其中包括 [widget](http://aralejs.org/widget/)、[base](http://aralejs.org/base/)、[class](http://aralejs.org/class/)、[events](http://aralejs.org/events/)、[attribute](http://aralejs.org/base/docs/attribute.html)、[aspect](http://aralejs.org/base/docs/aspect.html) 的属性和方法。


## 配置说明

### element `element|string`

弹出的浮层元素。也可以使用 template 参数传递模板来代替这个参数。

### trigger `element|string`

触发元素。

### triggerType `string`

触发类型，可选[hover|click|focus]，默认为 hover。

### delegateNode `element|string` `1.0.2+`

触发事件委托的对象。**当使用这个配置时，需要保证 trigger 参数是一个 selector !!!**

### effect `string`

基础的动画效果，可选[fade|slide]，默认为空 。fade 和 slide 可以并行。

### duration `number`

动画时长，默认 250 毫秒。

### delay `number`

延迟 hover 触发显示的时长，默认 70 毫秒。设为`-1`时，表示 popup 没有延迟显示和隐藏因素而变为 tooltip 的效果。

### disabled `boolean`

是否能触发弹出效果，可用`.set('disabled', true)`进行开关。


> 若有范例和演示中有其他参数（如自定义位置参数 align），请参见 Popup 的父类 [Overlay](http://aralejs.org/overlay/)。

## 属性

### activeTrigger

总是指向触发显示浮层的当前触发元素(jQuery Node 对象)。


## 事件

### animated

动画效果结束时的回调事件。`.on('animated', function(){ })`


## 最佳实践

```js
seajs.use(['arale/popup/{{版本号}}/popup'], function(Popup){
    var example = new Popup({
        trigger: '#triggerId',
        element: '#targetId'
    });
});
```

需要 tooltip 效果的请如下调用：

```js
seajs.use(['arale/popup/{{版本号}}/popup'], function(Popup){
    var example = new Popup({
        trigger: '#triggerId',
        element: '#targetId',
        delay: -1
    });
});
```

另外，Popup 已从 Overlay 继承了点击页面空白处浮层消失，以及窗口改变大小后浮层重新定位等被动属性，
详细请查看 [演示文档](http://aralejs.org/popup/examples/)。


### 多触发器的委托应用

实践中常常遇到多个触发元素对应一个弹出层，而且触发元素会动态添加的情况。
这样如果通过多个实例来绑定 popup 效果就会比较麻烦。

这时，可以使用 `delegateNode` 这个参数将事件绑定到委托节点上，配合多个 trigger 来实现。
在委托节点内动态添加了 trigger 后，也无须重新绑定。

```js
seajs.use(['popup'], function(Popup){
    new Popup({
        trigger: '.triggerClass',
        element: '#popup',
        delegateNode: '#delegateNode'
    });
});
```
[delegeteNode示例](examples/triggers.html#范例2-委托事件)

### 动态内容

当有多个触发元素，而弹出层是同一个，而且内容是根据不同触发元素而不同时，可以用下面的方式进行绑定：

```js
seajs.use(['popup'], function(Popup){
    new Popup({
        trigger: '.triggerClass', // 有多个触发元素
        element: '#popup'
    }).before('show', function() {
      // 通过 activeTrigger 拿到当前触发的触发元素
      var content = this.activeTrigger.html();
      // 动态设定内容
      this.element.html(content);
    });
});
```


HISTORY.md
# HISTORY

---

## 1.2.0

`tag:changed` 迁移到 spm3 。

## 1.1.6

`tag:improved` 升级 overlay 到 1.1.4

`tag:fixed` #23 修复一个浮层中有 select 控件时影响了 hover 时的隐藏行为的问题。

## 1.1.5

`tag:new`  新增 animated 动画结束的事件。

`tag:fixed` 修复一个动画效果显示元素时错误地调用了 hide 方法的问题。

`tag:fixed` 修复一个连续的动画效果显示问题。

`tag:changed`  去除了 click 链接时的阻止默认事件功能，现在需要自行处理。

## 1.1.4

`tag:fixed` #20 通过 template 生成的元素默认应该是不可见的。

`tag:fixed` 修复动画效果时在 ie6 下 iframe-shim 无法隐藏的问题。

`tag:improved` 升级 overlay 到 1.1.3


## 1.1.3

`tag:fixed` #19 修复一个 delegateNode 时导致点击 trigger 触发 blurHide 的问题。

## 1.1.2

`tag:improved` 升级 overlay 到 1.1.2

## 1.1.1

`tag:improved` 升级 overlay 到 1.1.1

`tag:fixed` 由于之前 build 出来的代码包含依赖两个不同版本的 position, 重新 build


## 1.1.0

`tag:improved` 对 arale/overlay 的依赖从 1.0.0 升级到 1.1.0 。

`tag:improved` 事件绑定使用 delegateEvents

## 1.0.2

`tag:improved` [#13](https://github.com/aralejs/popup/issues/13) 增加 delegateNode 参数，可以将触发事件委托到指定元素上。

## 1.0.1

`tag:fixed` 修复了一个 slide 动画效果导致定位错误的问题。

`tag:improved` 对 arale/overlay 的依赖升级为 1.0.1 。

## 1.0.0

`tag:changed` 对 arale/overlay 的依赖从 0.9.13 升级到 1.0.0 。

## 0.9.13

`tag:fixed` [#11](https://github.com/aralejs/popup/issues/11) delay 为 -1 时，未绑定 this.activeTrigger 而导致定位错误。

`tag:fixed` [#10](https://github.com/aralejs/popup/issues/10) this.activeTrigger 初始化时不是 jQuery 对象。

## 0.9.12

`tag:improved` delay 参数为 -1 时，表示 popup 没有延迟显示和隐藏因素而变为 tooltip 的效果。

`tag:unresolved` delay 为 -1 时，未绑定 this.activeTrigger 而导致定位错误。

`tag:fixed` [#8](https://github.com/aralejs/popup/issues/8) align 里的 baseElment 无法自定义的问题。


## 0.9.11

`tag:fixed` 修复多个实例公用一个浮出层时的动画重叠问题。

## 0.9.10

`tag:new` 增加基础的动画支持（fade | slide），以及动画时长 duration 参数。

`tag:fixed` 修复当 triggerType 为 focus 时无法操作弹出层上的内容的bug。

`tag:fixed` 修复一个动画为 slide 时的浮层定位错误。

`tag:fixed` [#6](https://github.com/aralejs/popup/issues/6) 直接调用 show 方法的定位错误。

`tag:improved` 依赖升级到 overlay 0.9.13。


## 0.9.9

`tag:new` 增加 disabled（是否能触发）参数。

`tag:fixed` 修复 trigger 为多个元素时的定位bug。

`tag:fixed` 修复 align.baseXY 为 [0, 0] 时无效的bug。

`tag:improved` 依赖升级到 overlay 0.9.12 ，将popup的显示时重新定位功能提到overlay中。

`tag:improved` delay 默认值调整为 70 毫秒。

`tag:improved` 代码优化重构。

## 0.9.8 

`tag:improved` 依赖升级到 overlay 0.9.10 版本，由此增加了窗口 resize 时，浮层重新定位的特性。

## 0.9.4 `2012-06-16`

`tag:changed` Triggerable 改名为 Popup。

## 0.9.3 `2012-06-05`

`tag:improved` 调整为 Triggerable 继承体系。

## 0.9.1 `2012-05-29`

`tag:changed` 按照新版 Widget 以及 Base 进行改造。

`tag:changed` 改成了继承 Overlay 的方式。

`tag:new` 支持 Delay 参数。


## 0.9.0 `2012-05-24`

`tag:new` 包含 Dropdown 基本功能。


tests/popup-spec.js
var $ = require('jquery');
var Popup = require('../src/popup');
var expect = require('expect.js');

describe('popup', function () {

  this.timeout(null);

  var template = '<div>' + '<a href="#" id="trigger1" class="trigger">popup</a>' + '<ul id="element1" style="display:none;">' + '<li>那些年，我们一起写过的单元测试...</li>' + '<li>卖萌是一种风格...</li>' + '</ul>' + '</div>';
  var pop;
  var element;

  beforeEach(function () {
    element = $(template).appendTo(document.body);
  });

  afterEach(function () {
    pop && pop.destroy();
    pop = null;
    element && element.remove();
  });

  it('instance', function () {
    pop = new Popup({
      trigger: '#trigger1',
      element: '#element1',
      effect: 'fade'
    });
    var trigger = pop.get('trigger');
    var align = pop.get('align');
    expect(trigger.attr('id')).to.be('trigger1');
    expect(pop.element.attr('id')).to.be('element1');
    expect(pop.get('triggerType')).to.be('hover');
    expect(align.baseXY).to.eql([0, '100%']);
    expect(align.selfXY).to.eql([0, 0]);
    expect(pop.get('effect')).to.be('fade');
    pop.set('align', null);
    expect(pop.get('align').baseElement[0].id).to.be('trigger1');
  });

  it('hover event', function (done) {
    var event1;
    var event2;
    pop = new Popup({
      trigger: '#trigger1',
      element: '#element1'
    });

    var showText = 'pop is shown';
    var hideText = 'pop is hidden';
    // 订阅事件
    pop.on('shown', function () {
      event1 = showText;
    }).on('hidden', function () {
      event2 = hideText;
    });
    // 发布事件
    pop.trigger('shown');
    pop.trigger('hidden');
    // 测试值
    expect(event1).to.be(showText);
    expect(event2).to.be(hideText);

    // 鼠标移入
    $('#trigger1').mouseover();
    setTimeout(function () {
      expect(pop.element.is(':visible')).to.be(true);

      // 鼠标移出                
      $('#trigger1').mouseout();

      setTimeout(function () {
        expect(pop.element.is(':visible')).to.be(false);
        done();
      }, 80);
    }, 80);

  });

  it('hover element make it visible', function (done) {
    pop = new Popup({
      trigger: '#trigger1',
      element: '#element1'
    });

    // 鼠标移入
    $('#trigger1').mouseover();
    setTimeout(function () {
      expect(pop.element.is(':visible')).to.be(true);

      $('#trigger1').mouseout();
      $('#element1').mouseover();

      setTimeout(function () {
        expect(pop.element.is(':visible')).to.be(true);

        $("#element1").mouseout();

        setTimeout(function () {
          expect(pop.element.is(':visible')).to.be(false);

          done();
        }, 80);
      }, 80);
    }, 80);
  });

  it('click event', function () {
    pop = new Popup({
      trigger: '#trigger1',
      element: '#element1',
      triggerType: 'click'
    });
    expect(pop.element.is(':visible')).to.be(false);
    $('#trigger1').click();
    expect(pop.element.is(':visible')).to.be(true);
    $('#trigger1').click();
    expect(pop.element.is(':visible')).to.be(false);
  });

  it('focus & blur event', function (done) {
    var input = $('<input type="text" />');
    input.appendTo(document.body);
    pop = new Popup({
      trigger: 'input',
      element: '#element1',
      triggerType: 'focus'
    });
    pop.render();
    expect(pop.element.is(':visible')).to.be(false);

    window.focus();
    input.focus();
    setTimeout(function () {
      expect(pop.element.is(':visible')).to.be(true);

      input.blur();
      setTimeout(function () {
        expect(pop.element.is(':visible')).to.be(false);

        input.remove();
        done();
      }, 100);
    }, 100);
  });

  it('blur when click element', function (done) {
    var input = $('<input type="text" />');
    input.appendTo(document.body);
    pop = new Popup({
      trigger: 'input',
      element: '#element1',
      triggerType: 'focus'
    });
    pop.render();
    expect(pop.element.is(':visible')).to.be(false);

    window.focus();
    input.focus();
    setTimeout(function () {
      expect(pop.element.is(':visible')).to.be(true);

      pop.element.mousedown();
      setTimeout(function () {
        expect(pop.element.is(':visible')).to.be(true);
        input.remove();
        done();
      }, 100);
    }, 100);
  });

  it('delay = -1', function () {
    pop = new Popup({
      trigger: '#trigger1',
      element: '#element1',
      delay: -1
    });
    pop.render();
    expect(pop.element.is(':visible')).to.be(false);
    $('#trigger1').trigger('mouseover');
    expect(pop.element.is(':visible')).to.be(true);
    $('#trigger1').trigger('mouseout');
    expect(pop.element.is(':visible')).to.be(false);
  });

  it('change align XY', function () {
    pop = new Popup({
      trigger: '#trigger1',
      element: '#element1',
      align: {
        baseXY: ['40%', 0],
        selfXY: [34, 120]
      }
    });
    var align = pop.get('align');
    expect(align.baseXY).to.eql(['40%', 0]);
    expect(align.selfXY).to.eql([34, 120]);
    pop.set('align', {
      baseXY: ['60%', -3],
      selfXY: [0, '20%']
    });
    align = pop.get('align');
    expect(align.baseXY).to.eql(['60%', -3]);
    expect(align.selfXY).to.eql([0, '20%']);
  });

  it('change align baseElement', function () {
    var pop = new Popup({
      trigger: '#trigger1',
      element: '#element1',
      align: {
        baseElement: 'body'
      }
    });
    var align = pop.get('align');
    expect(align.baseElement).to.eql('body');
  });

  it('change align baseElement after show', function () {
    var pop = new Popup({
      trigger: '#trigger1',
      element: '#element1',
      align: {
        baseElement: 'body'
      }
    });
    pop.show();
    var align = pop.get('align');
    expect(align.baseElement).to.eql('body');
  });

  it('change specified BaseElement', function () {
    var pop = new Popup({
      trigger: '#trigger1',
      element: '#element1'
    });
    expect(pop.get('align').baseElement[0].id).to.be('trigger1');
    expect(pop._specifiedBaseElement).to.be(undefined);
    pop.set('align', {
      baseElement: 'body'
    });
    expect(pop._specifiedBaseElement).to.be(true);
    expect(pop.get('align').baseElement).to.be('body');
  });

  it('disable', function () {
    pop = new Popup({
      trigger: '#trigger1',
      element: '#element1',
      triggerType: 'click',
      disabled: true
    });
    pop.render();
    expect(pop.element.is(':visible')).to.be(false);
    $('#trigger1').click();
    expect(pop.element.is(':visible')).to.be(false);
    pop.set('disabled', false);
    $('#trigger1').click();
    expect(pop.element.is(':visible')).to.be(true);
    $('#trigger1').click();
    expect(pop.element.is(':visible')).to.be(false);
  });

  it('delegate event', function (done) {
    pop = new Popup({
      trigger: '.trigger',
      element: '#element1',
      delegateNode: element
    });

    // 动态加入节点
    element.append('<a href="#" id="trigger2" class="trigger">popup</a>');
    // 鼠标移入
    $('#trigger2').mouseover();
    setTimeout(function () {
      expect(pop.element.is(':visible')).to.be(true);

      // 鼠标移出
      $('#trigger2').mouseout();

      setTimeout(function () {
        expect(pop.element.is(':visible')).to.be(false);
        done();
      }, 80);
    }, 80);

  });

  it('blur hide with delegate events', function () {
    pop = new Popup({
      trigger: '.trigger',
      element: '#element1',
      delegateNode: element,
      delay: -1
    });
    // 动态加入节点
    element.append('<a href="#" id="trigger2" class="trigger">popup</a>');
    expect(pop._relativeElements.length).to.be(2);
    // 鼠标移入
    $('#trigger2').mouseover();
    expect(pop._relativeElements.length).to.be(3);
    expect(pop.element.is(':visible')).to.be(true);
    $('#trigger2').click();
    expect(pop.element.is(':visible')).to.be(true);
    $('#element1').click();
    expect(pop.element.is(':visible')).to.be(true);
  });


  it('blur hide & triggers', function () {
    // 加入节点
    element.append('<a href="#" id="trigger2" class="trigger">popup</a>');

    pop = new Popup({
      trigger: '.trigger',
      triggerType: 'click',
      element: '#element1'
    });

    // 都不是激活状态
    expect(pop.element.is(':visible')).to.be(false);
    expect($('#trigger1')[0]._active).not.to.be.ok();
    expect($('#trigger2')[0]._active).not.to.be.ok();

    // 点击1，element 出现
    $('#trigger1').click();
    expect(pop.element.is(':visible')).to.be(true);
    expect($('#trigger1')[0]._active).to.be(true);
    expect($('#trigger2')[0]._active).to.be(false);

    // 点击2，element 出现
    $('#trigger2').click();
    expect(pop.element.is(':visible')).to.be(true);
    expect($('#trigger1')[0]._active).to.be(false);
    expect($('#trigger2')[0]._active).to.be(true);

    // 点击body，element 消失
    $('body').click();
    expect(pop.element.is(':visible')).to.be(false);
    expect($('#trigger1')[0]._active).to.be(false);
    expect($('#trigger2')[0]._active).to.be(false);

    // 点击2，element 再次出现
    $('#trigger2').click();
    expect(pop.element.is(':visible')).to.be(true);
    expect($('#trigger1')[0]._active).to.be(false);
    expect($('#trigger2')[0]._active).to.be(true);

    // 点击2，element 再次消失
    $('#trigger2').click();
    expect(pop.element.is(':visible')).to.be(false);
    expect($('#trigger1')[0]._active).to.be(false);
    expect($('#trigger2')[0]._active).to.be(false);

  });

  it('animate show & hide', function (done) {
    pop = new Popup({
      trigger: '#trigger1',
      element: '#element1',
      effect: 'fade',
      duration: 30
    });

    // 鼠标移入
    $('#trigger1').mouseover();
    setTimeout(function () {
      expect(pop.element.is(':visible')).to.be(true);

      // 鼠标移出
      $('#trigger1').mouseout();

      setTimeout(function () {
        expect(pop.element.is(':visible')).to.be(true);

        setTimeout(function () {
          expect(pop.element.is(':visible')).to.be(false);
          done();
        }, 80);
      }, 80);
    }, 80);

  });

  it('trigger selector', function () {
    var trigger = $('<div id="wrap"><a href="javascript:void(0)" class="trigger-element">Trigger</a></div>').appendTo(document.body);
    pop = new Popup({
      trigger: '#wrap .trigger-element',
      element: '#element1',
      triggerType: 'click'
    });
    $('#wrap .trigger-element').click();
    expect(pop.element.is(':visible')).to.be(true);
    $('#wrap .trigger-element').click();
    expect(pop.element.is(':visible')).to.be(false);
    trigger.remove();
  });

  it('should be visible when element is generated by template', function () {
    pop = new Popup({
      template: '<div>template</div>',
      trigger: '#trigger1',
      effect: 'slide'
    });
    pop.render();
    expect(pop.element.is(':visible')).to.be(false);
    pop.show();
    expect(pop.element.is(':visible')).to.be(true);
  });

  it('hide function should not be called when origin visible value is false', function () {
    var temp = 0;
    pop = new Popup({
      template: '<div>template</div>',
      trigger: '#trigger1',
      effect: 'fade',
      afterHide: function () {
        temp++;
      }
    });
    pop.render();
    expect(temp).to.be(0);
    pop.show();
    expect(temp).to.be(0);
    pop.hide();
    expect(temp).to.be(1);
  });

  it('animated event', function (done) {
    var temp = 0;
    pop = new Popup({
      template: '<div>template</div>',
      trigger: '#trigger1',
      effect: 'slide',
      duration: 80
    });
    pop.on('animated', function () {
      this.element.css('overflow', 'visible');
      temp++;
    });
    pop.show();
    expect(pop.element.css('overflow')).to.be('hidden');
    setTimeout(function () {
      expect(pop.element.css('overflow')).to.be('visible');
      expect(temp).to.be(1);
      done();
    }, 100);
  });

});


src/dropdown.js
define(function(require, exports, module) {

    var Popup = require('./popup');


    var Dropdown = Popup.extend({

        setup: function() {
            Dropdown.superclass.setup.call(this);
            this._tweakAlignDefaultValue();
        },
        
        show: function() {
            Dropdown.superclass.setup.call(this);
            this._setPosition();
        },

        // 调整 align 属性的默认值
        _tweakAlignDefaultValue: function() {
            var align = this.get('align');

            // 默认坐标在目标元素左下角
            if (align.baseXY.toString() === [0, 0].toString()) {
                align.baseXY = [0, '100%'];
            }

            // 默认基准定位元素为 trigger
            if (align.baseElement._id === 'VIEWPORT') {
                align.baseElement = this.get('trigger');
            }

            this.set('align', align);
        }
    });

    module.exports = Dropdown;

});


src/popup.js
var $ = require('jquery');
var Overlay = require('arale-overlay');


// Popup 是可触发 Overlay 型 UI 组件
var Popup = Overlay.extend({

  attrs: {
    // 触发元素
    trigger: {
      value: null,
      // required
      getter: function (val) {
        return $(val);
      }
    },

    // 触发类型
    triggerType: 'hover',
    // or click or focus
    // 触发事件委托的对象
    delegateNode: {
      value: null,
      getter: function (val) {
        return $(val);
      }
    },

    // 默认的定位参数
    align: {
      value: {
        baseXY: [0, '100%'],
        selfXY: [0, 0]
      },
      setter: function (val) {
        if (!val) {
          return;
        }
        if (val.baseElement) {
          this._specifiedBaseElement = true;
        } else if (this.activeTrigger) {
          // 若给的定位元素未指定基准元素
          // 就给一个...
          val.baseElement = this.activeTrigger;
        }
        return val;
      },
      getter: function (val) {
        // 若未指定基准元素，则按照当前的触发元素进行定位
        return $.extend({}, val, this._specifiedBaseElement ? {} : {
          baseElement: this.activeTrigger
        });
      }
    },

    // 延迟触发和隐藏时间
    delay: 70,

    // 是否能够触发
    // 可以通过set('disabled', true)关闭
    disabled: false,

    // 基本的动画效果，可选 fade|slide
    effect: '',

    // 动画的持续时间
    duration: 250

  },

  setup: function () {
    Popup.superclass.setup.call(this);
    this._bindTrigger();
    this._blurHide(this.get('trigger'));

    // 默认绑定activeTrigger为第一个元素
    // for https://github.com/aralejs/popup/issues/6
    this.activeTrigger = this.get('trigger').eq(0);

    // 当使用委托事件时，_blurHide 方法对于新添加的节点会失效
    // 这时需要重新绑定
    var that = this;
    if (this.get('delegateNode')) {
      this.before('show', function () {
        that._relativeElements = that.get('trigger');
        that._relativeElements.push(that.element);
      });
    }
  },

  render: function () {
    Popup.superclass.render.call(this);

    // 通过 template 生成的元素默认也应该是不可见的
    // 所以插入元素前强制隐藏元素，#20
    this.element.hide();
    return this;
  },

  show: function () {
    if (this.get('disabled')) {
      return;
    }
    return Popup.superclass.show.call(this);
  },

  // triggerShimSync 为 true 时
  // 表示什么都不做，只是触发 hide 的 before/after 绑定方法
  hide: function (triggerShimSync) {
    if (!triggerShimSync) {
      return Popup.superclass.hide.call(this);
    }
    return this;
  },

  _bindTrigger: function () {
    var triggerType = this.get('triggerType');

    if (triggerType === 'click') {
      this._bindClick();
    } else if (triggerType === 'focus') {
      this._bindFocus();
    } else {
      // 默认是 hover
      this._bindHover();
    }
  },

  _bindClick: function () {
    var that = this;

    bindEvent('click', this.get('trigger'), function (e) {
      // this._active 这个变量表明了当前触发元素是激活状态
      if (this._active === true) {
        that.hide();
      } else {
        // 将当前trigger标为激活状态
        makeActive(this);
        that.show();
      }
    }, this.get('delegateNode'), this);

    // 隐藏前清空激活状态
    this.before('hide', function () {
      makeActive();
    });

    // 处理所有trigger的激活状态
    // 若 trigger 为空，相当于清除所有元素的激活状态
    function makeActive(trigger) {
      if (that.get('disabled')) {
        return;
      }
      that.get('trigger').each(function (i, item) {
        if (trigger == item) {
          item._active = true;
          // 标识当前点击的元素
          that.activeTrigger = $(item);
        } else {
          item._active = false;
        }
      });
    }
  },

  _bindFocus: function () {
    var that = this;

    bindEvent('focus', this.get('trigger'), function () {
      // 标识当前点击的元素
      that.activeTrigger = $(this);
      that.show();
    }, this.get('delegateNode'), this);

    bindEvent('blur', this.get('trigger'), function () {
      var blurTrigger = this;
      setTimeout(function () {
        // 当 blur 的触发元素和当前的 activeTrigger 一样时才能干掉
        // 修复 https://github.com/aralejs/popup/issues/27
        if (!that._downOnElement && that.activeTrigger[0] === blurTrigger) {
          that.hide();
        }
        that._downOnElement = false;
      }, that.get('delay'));
    }, this.get('delegateNode'), this);

    // 为了当input blur时能够选择和操作弹出层上的内容
    this.delegateEvents("mousedown", function (e) {
      this._downOnElement = true;
    });
  },

  _bindHover: function () {
    var trigger = this.get('trigger');
    var delegateNode = this.get('delegateNode');
    var delay = this.get('delay');

    var showTimer, hideTimer;
    var that = this;

    // 当 delay 为负数时
    // popup 变成 tooltip 的效果
    if (delay < 0) {
      this._bindTooltip();
      return;
    }

    bindEvent('mouseenter', trigger, function () {
      clearTimeout(hideTimer);
      hideTimer = null;

      // 标识当前点击的元素
      that.activeTrigger = $(this);
      showTimer = setTimeout(function () {
        that.show();
      }, delay);
    }, delegateNode, this);

    bindEvent('mouseleave', trigger, leaveHandler, delegateNode, this);

    // 鼠标在悬浮层上时不消失
    this.delegateEvents("mouseenter", function () {
      clearTimeout(hideTimer);
    });
    this.delegateEvents("mouseleave", leaveHandler);

    this.element.on('mouseleave', 'select', function (e) {
      e.stopPropagation();
    });

    function leaveHandler(e) {
      clearTimeout(showTimer);
      showTimer = null;

      if (that.get('visible')) {
        hideTimer = setTimeout(function () {
          that.hide();
        }, delay);
      }
    }
  },

  _bindTooltip: function () {
    var trigger = this.get('trigger');
    var delegateNode = this.get('delegateNode');
    var that = this;

    bindEvent('mouseenter', trigger, function () {
      // 标识当前点击的元素
      that.activeTrigger = $(this);
      that.show();
    }, delegateNode, this);

    bindEvent('mouseleave', trigger, function () {
      that.hide();
    }, delegateNode, this);
  },

  _onRenderVisible: function (val, originVal) {
    // originVal 为 undefined 时不继续执行
    if (val === !!originVal) {
      return;
    }

    var fade = (this.get('effect').indexOf('fade') !== -1);
    var slide = (this.get('effect').indexOf('slide') !== -1);
    var animConfig = {};
    slide && (animConfig.height = (val ? 'show' : 'hide'));
    fade && (animConfig.opacity = (val ? 'show' : 'hide'));

    // 需要在回调时强制调一下 hide
    // 来触发 iframe-shim 的 sync 方法
    // 修复 ie6 下 shim 未隐藏的问题
    // visible 只有从 true 变为 false 时，才调用这个 hide
    var that = this;
    var hideComplete = val ?
    function () {
      that.trigger('animated');
    } : function () {
      // 参数 true 代表只是为了触发 shim 方法
      that.hide(true);
      that.trigger('animated');
    };

    if (fade || slide) {
      this.element.stop(true, true).animate(animConfig, this.get('duration'), hideComplete).css({
        'visibility': 'visible'
      });
    } else {
      this.element[val ? 'show' : 'hide']();
    }
  }

});

module.exports = Popup;

// 一个绑定事件的简单封装
function bindEvent(type, element, fn, delegateNode, context) {
  var hasDelegateNode = delegateNode && delegateNode[0];

  context.delegateEvents(
  hasDelegateNode ? delegateNode : element, hasDelegateNode ? type + " " + element.selector : type, function (e) {
    fn.call(e.currentTarget, e);
  });
}
